### PR TITLE
Added resource owner password credentials grant flow

### PR DESCRIPTION
**Pull request by letrungkien0210**
Hi @jakubknejzlik ,

Please review.

Thanks

### CODE REVIEW

File: src/login.js [hunk 1]
```diff
@@ -23,23 +24,59 @@ let providers = {
...
+  
+  handleSubmit(e){
+    e.preventDefault();
+    const options = {
+      uri: this.props.apiAuthorization,
+      qs: {
+        grant_type: "password",
+        username: this.state.username,
+        password: this.state.password,
+        client_id: "web"
```
Comment (jakubknejzlik) [inline src/login.js]: Client_id should be configurable by props

File: src/login.js [hunk 2]
```diff
@@ -23,23 +24,59 @@ let providers = {
...
+      },
+      json: true
+    }
+    rp(options)
+      .then(token => {
+        localStorage.set("token", token);
+        if(this.props.redirectUrl){
+          window.location.href = this.props.redirectUrl;
```
Comment (jakubknejzlik) [inline src/login.js]: We should redirect with access_token in querystring

File: src/login.js [hunk 3]
```diff
@@ -23,23 +24,59 @@ let providers = {
...
+        }
+      })
+      .catch( error => {
+        console.error(error);
```
Comment (jakubknejzlik) [inline src/login.js]: Can we display the error instead of logging it to console?

File: src/login.js [hunk 4]
```diff
@@ -130,7 +167,9 @@ Login.defaultProps = {
...
   userCredentialsEnabled: true,
-  authProviders: ["facebook", "google", "gitlab"]
-};
+  authProviders: ["facebook", "google", "gitlab"],
+  apiAuthorization: "",
+  redirectUrl: ""
+}
 
-export default Login;
+export default  Login;
```
Comment (jakubknejzlik) [inline src/login.js]: What's the added space for?

File: src/login.js [hunk 5]
```diff
@@ -23,23 +24,59 @@ let providers = {
...
+    const name = target.name;
+    this.setState({
+      [name]: target.value
+    });
+  }
+  
+  handleSubmit(e){
+    e.preventDefault();
+    const options = {
+      uri: this.props.apiAuthorization,
```
Comment (jakubknejzlik) [inline src/login.js]: Please rename to authUrl

**Review by jakubknejzlik — CHANGES_REQUESTED** [review]

File: src/login.js (hunk 5 above)
Comment (letrungkien0210) [reply src/login.js]: uri is the parameter of request-promise, I don't think we can rename it.

**Comment by letrungkien0210** [comment]
Hi @jakubknejzlik ,

Please review.

Thanks.

File: src/login.js (hunk 5 above)
Comment (jakubknejzlik) [reply src/login.js]: I meant the props key :). So: this.props.authUrl

Also with password grant type we use token endpoint (in props we will have both authUrl and tokenUrl, but password use only tokenUrl)

Comment (jakubknejzlik) [reply src/login.js]: Also the data is sent by POST method

File: src/login.js [hunk 6]
```diff
@@ -23,23 +24,59 @@ let providers = {
...
+      qs: {
```
Comment (jakubknejzlik) [inline src/login.js]: According to https://tools.ietf.org/html/rfc6749#section-4.3.2 data should be sent by POST method in `application/x-www-form-urlencoded`. Also scope attribute/property is missing.

**Review by jakubknejzlik — CHANGES_REQUESTED** [review]

File: src/login.js (hunk 6 above)
Comment (letrungkien0210) [reply src/login.js]: Base on section-4.3.2, I also removed client_id.

**Comment by letrungkien0210** [comment]
Hi @jakubknejzlik ,

I fixed base on your comments.

Please review.

**Pull request merged**